mmcls/models/backbones/levit.py

```diff
@@ -0,0 +1,586 @@
+            torch.zeros(num_heads, len(attention_offsets)))
+        self.register_buffer('attention_bias_idxs',
+                             torch.LongTensor(idxs).view(N_, N))
+
+    @torch.no_grad()
+    def train(self, mode=True):
+        super(AttentionSubsample, self).train(mode)
+        if mode and hasattr(self, 'ab'):
+            del self.ab
+        else:
+            self.ab = self.attention_biases[:, self.attention_bias_idxs]
+
+    def forward(self, x):
+        B, N, C = x.shape
+        k, v = self.kv(x).view(B, N, self.num_heads,
+                               -1).split([self.key_dim, self.d], dim=3)
+        k = k.permute(0, 2, 1, 3)  # BHNC
+        v = v.permute(0, 2, 1, 3)  # BHNC
+        q = self.q(x).view(B, self.resolution_2, self.num_heads,
+                           self.key_dim).permute(0, 2, 1, 3)
+
+        attn = (q @ k.transpose(-2, -1)) * self.scale + \
+               (self.attention_biases[:, self.attention_bias_idxs]
+                if self.training else self.ab)
+        attn = attn.softmax(dim=-1)
+
+        x = (attn @ v).transpose(1, 2).reshape(B, -1, self.dh)
+        x = self.proj(x)
+        return x
+
+
+@MODELS.register_module()
+class LeViT(BaseBackbone):
+    """Vision Transformer with support for patch or hybrid CNN input stage."""
+
+    def __init__(self,
+                 img_size=224,
+                 patch_size=16,
+                 embed_dim=[192],
+                 key_dim=[64],
+                 depth=[12],
+                 num_heads=[3],
+                 attn_ratio=[2],
+                 mlp_ratio=[2],
+                 hybrid_backbone=None,
+                 down_ops=[],
+                 attention_activation=torch.nn.Hardswish,
+                 mlp_activation=torch.nn.Hardswish,
+                 drop_path=0,
+                 out_indices=(2, ),
+                 deploy=False):
+        super(LeViT, self).__init__()
+
+        self.out_indices = out_indices
+
+        self.num_features = embed_dim[-1]
+        self.embed_dim = embed_dim
+        if not hybrid_backbone:
+            hybrid_backbone = hybrid_cnn(embed_dim[0])
+        self.patch_embed = hybrid_backbone
+        self.blocks = [[]]
+        self.size = []
+        self.deploy = deploy
+
+        down_ops.append([''])
+        resolution = img_size // patch_size
+        for i, (ed, kd, dpth, nh, ar, mr, do) in enumerate(
+                zip(embed_dim, key_dim, depth, num_heads, attn_ratio,
+                    mlp_ratio, down_ops)):
+            for _ in range(dpth):
+                self.blocks[-1].append(
+                    Residual(
+                        Attention(
+                            ed,
+                            kd,
+                            nh,
+                            attn_ratio=ar,
+                            activation=attention_activation,
+                            resolution=resolution,
+                        ), drop_path))
+                if mr > 0:
```

Comment:
This condition is useless if `mlp_ratio` won't equal zero.

Reply:
Sorry, I don't understand what you mean. Only when mlp_ MLP is meaningful only when the ratio is greater than 0

Reply:
Sorry, I don't understand what you mean. Only when the mlp_ ratio is greater than 0 MLP is meaningful